### PR TITLE
Minor: ScanExec code cleanup and additional documentation

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

N/A

## Rationale for this change

I recently learned more about how ScanExec works so I thought I would add some documentation to help other contributors.

## What changes are included in this PR?

- More documentation
- Some pedantic code cleanup changes
- Use a single metrics timer instead of two timers when fetching batches
- Create a vector with the correct capacity to avoid overhead of reallocation
- Re-use CastOptions across all batches instead of creating per-batch

## How are these changes tested?

Existing tests

### CODE REVIEW

**Review by huaxingao — APPROVED** [review]
LGTM. Thanks for the PR @andygrove

File: native/core/src/execution/operators/scan.rs [hunk 1]
```diff
@@ -109,11 +120,15 @@ impl ScanExec {
...
     ///
-    /// TODO: revisit this once DF has imprved its dictionary type support. Ideally we shouldn't
+    /// TODO: revisit this once DF has improved its dictionary type support. Ideally we shouldn't
     ///   do this in Comet but rather let DF to handle it for us.
     fn unpack_dictionary_type(dt: &DataType) -> DataType {
         if let DataType::Dictionary(_, vt) = dt {
-            if !matches!(vt.as_ref(), DataType::Utf8 | DataType::Binary) {
+            if !matches!(
+                vt.as_ref(),
+                DataType::Utf8 | DataType::LargeUtf8 | DataType::Binary
```
Comment (viirya) [inline native/core/src/execution/operators/scan.rs:129]: How about LargeBinary?

Comment (andygrove) [reply native/core/src/execution/operators/scan.rs]: Added. Thanks.

**Review by viirya — APPROVED** [review]

**Review by parthchandra — APPROVED** [review]
lgtm. 
Thank you for the additional documentation!

**Pull request merged**